flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2619,157 +2832,541 @@ class ArrayExprLowering {
-      for (const auto &s : shape)
-        if (!s || !Fortran::evaluate::IsConstantExpr(*s))
-          return false;
-      return true;
-    };
-    if (destShape.has_value() && destShapeIsConstant(*destShape)) {
-      auto idxTy = builder.getIndexType();
-      llvm::SmallVector<mlir::Value> shape;
-      for (const auto &s : *destShape)
-        shape.emplace_back(builder.createConvert(
-            loc, idxTy, convertOptExtentExpr(converter, stmtCtx, s)));
-      return shape;
-    }
+    if (explicitSpace)
+      return explicitImpliedShape;
+    // Use the precomputed destination shape.
+    if (!destShape.empty())
+      return destShape;
+    // Otherwise, use the destination's shape.
     if (destination)
       return getShape(destination);
+    // Otherwise, use the first ArrayLoad operand shape.
     if (!arrayOperandLoads.empty())
       return getShape(arrayOperandLoads[0]);
-    assert(destinationMutableBox && "shape must have been deduced if this is "
-                                    "not an allocatable assignment");
-    return {};
+    fir::emitFatalError(getLoc(),
+                        "failed to compute the array expression shape");
   }
 
-  /// Build the iteration space into which the array expression will be lowered.
-  std::pair<IterationSpace, mlir::OpBuilder::InsertPoint>
-  genIterSpace(mlir::Type resultType) {
+  /// Returns the type of the outermost projection.
+  mlir::Type explicitShapeResultType(
+      const Fortran::evaluate::Expr<Fortran::evaluate::SomeType> &exp) {
+    TODO(getLoc(), "determine the result type");
+  }
+
+  /// Force the binding of an explicit symbol. This is used to bind and re-bind
+  /// a concurrent control symbol to its value.
+  void forceControlVariableBinding(mlir::Location loc,
+                                   const Fortran::semantics::Symbol *sym,
+                                   mlir::Value inducVar) {
+    assert(sym && "There must be a symbol to bind");
+    auto toTy = converter.genType(Fortran::evaluate::SymbolRef(*sym));
+    // FIXME: this should be a "per iteration" temporary.
+    auto tmp = builder.createTemporary(
+        loc, toTy, toStringRef(sym->name()),
+        llvm::ArrayRef<mlir::NamedAttribute>{getAdaptToByRefAttr()});
+    auto cast = builder.createConvert(loc, toTy, inducVar);
+    builder.create<fir::StoreOp>(loc, cast, tmp);
+    symMap.addSymbol(*sym, tmp, /*force=*/true);
+  }
+
+  /// Lower the mask expression `e` associated with the concurrent control
+  /// header. A concurrent control header has 0 or 1 mask expressions.
+  ///
+  /// The algorithm is:
+  /// 1. Lower the explicit loop nest and conditionalize along the way.
+  /// 2. Compute the mask values and store them into a compressed temporary.
+  /// 3. Return the temporary array and the shape op that describes it.
+  Fortran::lower::MaskAddrAndShape
+  genExplicitMaskExpr(Fortran::lower::FrontEndExpr fex) {
+    using FESymbol = Fortran::lower::IterationSpaceExpr::FrontEndSymbol;
```

Comment:
Are you entirely sure that these expression have to be pure ? The standard states in 10.1.4 point 2 that:

> Execution of a function reference in [...] the concurrent-limits and concurrent-steps in a FORALL statement (10.2.4) is permitted to define variables in the subsidiary [...] forall-assignment-stmt.

That being said, if the FORALL concurrent-limits and concurrent-steps may depend on higher level FORALL indices, it sounds hard to ensure they are evaluated only once... So it may be worth studying a bit what is the behavior of existing compilers here before starting to do more complex things.